lib/serum/build/file_loader/templates.ex

```diff
@@ -11,35 +11,67 @@ defmodule Serum.Build.FileLoader.Templates do
   alias Serum.Result
   alias Serum.Theme
 
+  @required_templates MapSet.new(~w(base list page post))
+
   @doc false
   @spec load(binary()) :: Result.t([Serum.File.t()])
   def load(src) do
     put_msg(:info, "Loading template files...")
 
-    case Theme.get_templates() do
-      {:ok, paths} ->
-        paths
-        |> Map.merge(get_project_templates(src), fn _, v1, v2 ->
-          (File.exists?(v2) && v2) || v1
-        end)
-        |> Enum.map(&elem(&1, 1))
-        |> Plugin.reading_templates()
-        |> case do
-          {:ok, files} -> read_files(files)
-          {:error, _} = plugin_error -> plugin_error
-        end
-
+    with {:ok, theme_paths} <- Theme.get_templates(),
+         {:ok, proj_paths} <- get_project_templates(src),
+         merged <- Map.merge(theme_paths, proj_paths, &merge_fun/3),
+         :ok <- validate_required(merged, src) do
+      merged
+      |> Enum.map(&elem(&1, 1))
+      |> Plugin.reading_templates()
+      |> case do
+        {:ok, files} -> read_files(files)
+        {:error, _} = plugin_error -> plugin_error
+      end
+    else
       {:error, _} = error ->
         error
     end
   end
 
-  @spec get_project_templates(binary()) :: map()
+  @spec get_project_templates(binary()) :: Result.t(map())
   defp get_project_templates(src) do
     templates_dir = get_subdir(src, "templates")
 
-    ["base", "list", "page", "post"]
-    |> Enum.map(&{&1, Path.join(templates_dir, &1 <> ".html.eex")})
-    |> Map.new()
+    map =
+      templates_dir
+      |> Path.join("*.html.eex")
+      |> Path.wildcard()
+      |> Enum.map(&{Path.basename(&1, ".html.eex"), &1})
+      |> Map.new()
+
+    {:ok, map}
+  end
+
+  @spec merge_fun(binary(), binary(), binary()) :: binary()
+  defp merge_fun(_k, from_theme, from_proj) do
+    (File.exists?(from_proj) && from_proj) || from_theme
+  end
+
+  @spec validate_required(map(), binary()) :: Result.t()
+  defp validate_required(map, src) do
+    existing_templates = map |> Map.keys() |> MapSet.new()
+
+    @required_templates
+    |> MapSet.difference(existing_templates)
+    |> MapSet.to_list()
+    |> case do
+      [] ->
+        :ok
+
+      missings when is_list(missings) ->
+        errors =
+          Enum.map(missings, fn missing ->
+            {:error, {:enoent, Path.join([src, "templates", missing]), 0}}
+          end)
```

Comment:
doesn't this create a regression here? https://github.com/Dalgona/Serum/issues/40

Reply:
Oh, please. I'm tracking

Reply:
after finishing my cup of coffee, i think i have to disagree with myself

Reply:
You don't have to do that. I rather apologize for such an unprofessional response. 🙇 Let me give you an excuse, I couldn't sleep overnight and I'm feeling kinda exhausted right now.

Since I cannot access my dev machine right now, I will take a look at it later this evening (KST). Thank you.

Reply:
I don't think this causes a regression. This code just checks if all four required templates, `base`, `list`, `page`, and `post`, are available for use. The same check has also been performed before this PR, but in different (somewhat indirect) way.

If you don't have any blog post in your project, Serum still won't generate any post and list of posts. In that case, `list.html.eex` and `post.html.eex` should still be present either in your project or a theme, but they will never be used. (Currently, I don't have any plan to remove this requirement. It's just because of my personal preference, though.)

Reply:
Please hit the "Resolve conversation" button if you agree with my comment above, or feel free to leave more comments for further discussion. Thanks.

Reply:
there's no resolve button here, since my comment came after merge

but I'll gladly test this with my site for you, when I'm back on my laptop

Reply:
everything works perfectly!

Reply:
Thanks!